symengine/mul.cpp

```diff
@@ -456,10 +456,15 @@ RCP<const Basic> Mul::subs(const map_basic_basic &subs_dict) const
     RCP<const Number> coef = coef_;
     map_basic_basic d;
     for (const auto &p: dict_) {
-        RCP<const Basic> factor_old = Mul::from_dict(one, {{p.first, p.second}});
+        RCP<const Basic> factor_old;
+        if (eq(*p.second, *one)) {
+            factor_old = p.first;
+        } else {
+            factor_old = make_rcp<Pow>(p.first, p.second);
+        }
         RCP<const Basic> factor = factor_old->subs(subs_dict);
         if (factor == factor_old) {
-            Mul::dict_add_term_new(outArg(coef), d, p.second, p.first);
+            Mul::dict_add_term(d, p.second, p.first);
```

Comment:
This seems unrelated to changing the integer class, or is it related? What was the motivation for this change? I am a little worried that changes like these might break things, and since this PR is so big, we don't want to be messing with such things, unless we have to.

If this is a good, but unrelated fix, then it should go into a separate PR, so that it's easy to revert if needed.